Support compiling with `-fvisibility=hidden`

Split into:
* [x] #442
* [ ] #443

Compiling with `-fvisibility=hidden` gives the following error which @kkysen hit with dav1d. This is caused because the undefined references become local symbols in the other .so so we need explicit visibility attributes on those variables to avoid this.
```
[7/7] Linking C executable tests/two_keys_minimal/two_keys_minimal
FAILED: tests/two_keys_minimal/two_keys_minimal
cd /home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal && /usr/bin/objcopy --redefine-syms=/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/two_keys_minimal_call_gates_1.objcopy /home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/./main.c.o && cd /home/ayrton/immunant/IA2-Phase2/build && /usr/bin/cc  -fsanitize=undefined -Wl,--export-dynamic -Wl,@/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal/two_keys_minimal_call_gates_1.ld -Wl,--print-map-locals -pthread -Wl,--wrap=pthread_create -Wl,--wrap=main -Wl,-z,now -Wl,-z,relro -Wl,-T/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2/padding.ld -Wl,--dynamic-list=/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2/dynsym.syms -Wl,-wrap,calloc -Wl,-wrap,free -Wl,-wrap,malloc -Wl,-wrap,memalign -Wl,-wrap,posix_memalign -Wl,-wrap,pvalloc -Wl,-wrap,realloc -Wl,-wrap,valloc -Wl,-wrap,malloc_usable_size -Wl,-wrap,realpath -Wl,-wrap,strdup -Wl,-wrap,strndup -Wl,-wrap,getcwd -Wl,-wrap,asprintf -Wl,-wrap,vasprintf tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o -o tests/two_keys_minimal/two_keys_minimal  -Wl,-rpath,/home/ayrton/immunant/IA2-Phase2/build/runtime/partition-alloc:/home/ayrton/immunant/IA2-Phase2/build/tests/two_keys_minimal:/home/ayrton/immunant/IA2-Phase2/build/runtime/libia2  -ldl  runtime/libia2/liblibia2.a  runtime/partition-alloc/libpartition-alloc.so  -lcriterion  tests/two_keys_minimal/libtwo_keys_minimal_lib.so  tests/two_keys_minimal/libtwo_keys_minimal_call_gates.so  runtime/libia2/liblibia2.a  -ldl  runtime/libia2/libc.so.6 && :
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `ia2_stackptr_for_pkru':
main.c:(.text+0x24e): undefined reference to `ia2_stackptr_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `init_stacks_and_setup_tls':
main.c:(.text+0x2bf): undefined reference to `init_tls_2'
/usr/bin/ld: main.c:(.text+0x2d0): undefined reference to `ia2_stackptr_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `ia2_init':
main.c:(.text+0x395): undefined reference to `ia2_setup_destructors_2'
/usr/bin/ld: tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o: in function `print_message':
main.c:(.text+0x17fd): undefined reference to `plugin_secret'
/usr/bin/ld: tests/two_keys_minimal/two_keys_minimal: hidden symbol `two_keys_plugin_impl' in tests/two_keys_minimal/CMakeFiles/two_keys_minimal.dir/main.c.o is referenced by DSO
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

--- a/tools/rewriter/GenCallAsm.cpp
+++ b/tools/rewriter/GenCallAsm.cpp
@@ -979,7 +979,7 @@ std::string emit_asm_wrapper(AbiSignature &sig,
 
   add_comment_line(aw, "Wrapper for "s + sig_string(sig, target_name) + ":");
   add_asm_line(aw, ".text");
-  if (!as_macro) {
+  if (kind != WrapperKind::PointerToStatic) {
     add_asm_line(aw, ".global "s + wrapper_name);
   } else {
     add_asm_line(aw, ".local "s + wrapper_name);

--- a/tools/rewriter/GenCallAsm.h
+++ b/tools/rewriter/GenCallAsm.h
@@ -10,6 +10,8 @@ enum class WrapperKind {
   Direct,
   // Indirect call through a pointer sent to another compartment
   Pointer,
+  // Indirect call through a pointer sent to another compartment
+  PointerToStatic,
   // Indirect call through a pointer received from another compartment
   IndirectCallsite,
 };

--- a/tools/rewriter/SourceRewriter.cpp
+++ b/tools/rewriter/SourceRewriter.cpp
@@ -1457,7 +1457,11 @@ int main(int argc, const char **argv) {
       macros_defining_wrappers += asm_wrapper;
       macros_defining_wrappers += ");\n";
 
-      /* Invoke the macro we just defined in the source file defining the target function */
+      /*
+       * Invoke IA2_DEFINE_WRAPPER from ia2.h in the source file defining the
+       * target function. This expands to the IA2_DEFINE_WRAPPER_* macro we just
+       * defined
+       */
       auto filename = fn_decl_pass.fn_definitions[fn_name];
       std::ofstream source_file(filename, std::ios::app);
       source_file << "IA2_DEFINE_WRAPPER(" << fn_name << ")\n";
@@ -1498,6 +1502,11 @@ int main(int argc, const char **argv) {
 
         header_out << "extern " << opaque << " " << wrapper_name << ";\n";
 
+        /*
+         * Invoke IA2_DEFINE_WRAPPER from ia2.h in the source file defining the
+         * target function. This expands to the IA2_DEFINE_WRAPPER_* macro we just
+         * defined
+         */
         source_file << "IA2_DEFINE_WRAPPER(" << fn_name << ")\n";
       }
     }

--- a/tools/rewriter/SourceRewriter.cpp
+++ b/tools/rewriter/SourceRewriter.cpp
@@ -911,6 +911,8 @@ public:
     if (definition) {
       defined_fns[pkey].insert(fn_name);
       fn_pkeys[fn_name] = pkey;
+      Filename filename = get_expansion_filename(fn_node->getLocation(), sm);
+      fn_definitions[fn_name] = filename;
     } else {
       declared_fns[pkey].insert(fn_name);
     }
@@ -920,6 +922,7 @@ public:
   std::set<Function> declared_fns[MAX_PKEYS];
   std::map<Function, AbiSignature> abi_signatures;
   std::map<Function, Pkey> fn_pkeys;
+  std::map<Function, Filename> fn_definitions;
 };
 
 static void create_file(llvm::raw_fd_ostream *file[MAX_PKEYS], int i, const char *extension) {
@@ -1418,8 +1421,20 @@ int main(int argc, const char **argv) {
   }
 
   std::cout << "Generating function pointer wrappers\n";
+  std::string macros_defining_wrappers;
+  /*
+   * This loops over all non-static address-taken functions but we only want to
+   * define one call gate for each so we need to track them in case a function
+   * has its address taken in multiple places
+   */
+  std::set<Function> generated_wrappers = {};
   // Define wrappers for function pointers (i.e. those referenced by IA2_FN)
   for (const auto &[fn_name, opaque] : ptr_expr_pass.addr_taken_fns) {
+    if (generated_wrappers.contains(fn_name)) {
+        continue;
+    }
+    llvm::errs() << " inserting " << fn_name << " into set\n";
+    generated_wrappers.insert(fn_name);
     /*
      * Declare these wrapper in the output header so that IA2_FN can reference
      * them. e.g. extern struct IA2_fnptr_ZTSFiiE __ia2_foo;
@@ -1435,10 +1450,18 @@ int main(int argc, const char **argv) {
       AbiSignature c_abi_sig = fn_decl_pass.abi_signatures[fn_name];
       std::string asm_wrapper =
           emit_asm_wrapper(c_abi_sig, wrapper_name, fn_name,
-                           WrapperKind::Pointer, 0, target_pkey, Target);
-      wrapper_out << "asm(\n";
-      wrapper_out << asm_wrapper;
-      wrapper_out << ");\n";
+                           WrapperKind::Pointer, 0, target_pkey, Target,
+                           true /* as_macro */);
+      macros_defining_wrappers += "#define IA2_DEFINE_WRAPPER_"s + fn_name + " \\\n";
+      macros_defining_wrappers += "asm(\\\n";
+      macros_defining_wrappers += asm_wrapper;
+      macros_defining_wrappers += ");\n";
+
+      /* Invoke the macro we just defined in the source file defining the target function */
+      auto filename = fn_decl_pass.fn_definitions[fn_name];
+      std::ofstream source_file(filename, std::ios::app);
+      source_file << "IA2_DEFINE_WRAPPER(" << fn_name << ")\n";
+
     } else {
       header_out << "asm(\n";
       header_out << "  \".set " << wrapper_name << ", __real_" << fn_name << "\\n\"\n";
@@ -1449,7 +1472,6 @@ int main(int argc, const char **argv) {
   std::cout << "Generating function pointer wrappers for static functions\n";
   // Define wrappers for pointers to static functions (also those referenced by
   // IA2_FN)
-  std::string static_wrappers;
   for (const auto &[filename, addr_taken_fns] :
        ptr_expr_pass.internal_addr_taken_fns) {
 
@@ -1467,12 +1489,12 @@ int main(int argc, const char **argv) {
         AbiSignature c_abi_sig = fn_decl_pass.abi_signatures[fn_name];
 
         std::string asm_wrapper = emit_asm_wrapper(
-            c_abi_sig, wrapper_name, fn_name, WrapperKind::Pointer, 0,
+            c_abi_sig, wrapper_name, fn_name, WrapperKind::PointerToStatic, 0,
             target_pkey, Target, true /* as_macro */);
-        static_wrappers += "#define IA2_DEFINE_WRAPPER_"s + fn_name + " \\\n";
-        static_wrappers += "asm(\\\n";
-        static_wrappers += asm_wrapper;
-        static_wrappers += ");\n";
+        macros_defining_wrappers += "#define IA2_DEFINE_WRAPPER_"s + fn_name + " \\\n";
+        macros_defining_wrappers += "asm(\\\n";
+        macros_defining_wrappers += asm_wrapper;
+        macros_defining_wrappers += ");\n";
 
         header_out << "extern " << opaque << " " << wrapper_name << ";\n";
 
@@ -1490,7 +1512,7 @@ int main(int argc, const char **argv) {
   }
   header_out << "asm(\"__libia2_abort:\\n\"\n"
              << "    \"" << undef_insn << "\");\n";
-  header_out << static_wrappers.c_str();
+  header_out << macros_defining_wrappers.c_str();
 
   for (int i = 0; i < num_pkeys; i++) {
     if (ld_args_out[i] != nullptr) {

